src/docfx/build/toc/TableOfContentsLoader.cs

```diff
@@ -79,6 +82,11 @@ internal class TableOfContentsLoader
 
                 var node = _parser.Parse(file.FilePath, errors);
                 node.Items = LoadTocNodes(node.Items, file, rootPath, referencedFiles, referencedTocs, errors);
+
+                if (file == rootPath)
+                {
+                    _contentValidator.ValidateTocEntryDuplicated(file, node);
```

Comment:
The `referencedFiles` property contains all the markdown files referenced by this TOC, could use that over traversing the TOC nodes again. And also can change the signature to `List<SourceInfo<Document>>` (and also update insertion logic) to retrieve exactly where the source file is referenced in this TOC file.

Reply:
This is much better, thanks for the tip. I decided not to add the source info because there could potentially be multiple nodes in the same file that could be duplicated and the source info wouldn't be of much help in that case. The node file names are output in the error for authors to find and fix the problem(s).